core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -159,6 +159,10 @@ public TableScan option(String property, String value) {
 
   @Override
   public TableScan project(Schema projectedSchema) {
+    if (context.selectedColumns() != null) {
+      throw new IllegalStateException("Cannot project schema when selected columns is specified.");
+    }
```

Comment:
We would normally use a `Precondition` check here:

```java
Preconditions.checkState(context.selectedColumns() == null, "Cannot ...");
```

I was also going to say that I think we should check the reverse as well, in case `project` is called before `select`, but it looks like the projection is tracked outside of the context for some reason. I don't think that's quite right. We probably want to move `project` to the context as well.

@edgarRd, do you know why `project` doesn't use the scan context?

Reply:
Another thing is: `BaseTableScan.lazyColumnProjection` looks weird. It looks only work for `DataTableScan`. Because other table scan like `MetadataTableScan`, the file/expected scan schema is different from `table.schema()`. But in `BaseTableScan.lazyColumnProjection`, it use `table.schema()` to select. And for `rowFilter`, it looks only works for `DataTableScan` selection, should not affect `MetadataTableScan`s selection.

In other words, we may need to implement different selections for various scans.

Reply:
This was the idea on not putting schema in the context: https://github.com/apache/iceberg/pull/1115#discussion_r441068118 but I guess it could be part of the context, it would need to be initialized and in this case another context would need to be created with the new schema from the projection.

Reply:
Thanks @edgarRd! I think it makes sense to move that into the context and initialize it as null. If the projection is null, then use the column list. If not, then use the projection. And then both methods can validate that the other has not been called.

I think it would be okay to do this even if it only works for `DataTableScan`. If the other scans don't support it, then they should throw exceptions instead of allowing the projection to be set.

Reply:
Yeah, that makes sense @rdblue. Should we have another PR for that change?

Reply:
Hi @edgarRd , If you don't mind, I think I can finish it in this PR.